makeflow/src/makeflow_wrapper_enforcement.c

```diff
@@ -0,0 +1,140 @@
+	w->command = xxstrdup(enforcer_pattern "%%");
+}
+
+char *makeflow_wrap_enforcer( char *result, struct dag_node *n, struct makeflow_wrapper *w, struct list *input_list, struct list *output_list )
+{
+	if(!w) return result;
+
+	struct list *enforcer_paths = list_create();
+	struct dag_file *f;
+	FILE *enforcer;
+	char *enforcer_path = string_format(enforcer_pattern "%d", n->nodeid);
+	char *mountlist_path = string_format(mountlist_pattern "%d", n->nodeid);
+	char *tmp_path = string_format(tmp_pattern "%d", n->nodeid);
+	char *vartmp_path = string_format(vartmp_pattern "%d", n->nodeid);
+	char *shm_path = string_format(shm_pattern "%d", n->nodeid);
+
+	/* Create local directories to bind to /tmp, /var/tmp, and /dev/shm */
+	if (mkdir(tmp_path, S_IRWXU) || mkdir(vartmp_path, S_IRWXU) || mkdir(shm_path, S_IRWXU)) {
+		fatal("could not create temp directory: %s", strerror(errno));
+	}
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, tmp_path));
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, vartmp_path));
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, shm_path));
+
+	/* make an invalid mountfile to send */
+	int mountlist_fd = open(mountlist_path, O_WRONLY|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR);
+	if (mountlist_fd == -1) {
+		fatal("could not create `%s': %s", mountlist_path, strerror(errno));
+	}
+	write(mountlist_fd, "mountlist\n", 10);
+	close(mountlist_fd);
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, mountlist_path));
+
+	/* and generate a wrapper script with the current nodeid */
+	int enforcer_fd = open(enforcer_path, O_WRONLY|O_CREAT|O_EXCL, S_IRWXU);
+	if (enforcer_fd == -1 || (enforcer = fdopen(enforcer_fd, "w")) == NULL) {
+		fatal("could not create `%s': %s", enforcer_path, strerror(errno));
+	}
+	fchmod(enforcer_fd, 0755);
+	fprintf(enforcer, "#!/bin/sh\n\n");
+	fprintf(enforcer, "MOUNTFILE='%s'\n", mountlist_path);
+	fprintf(enforcer, "cat > \"$MOUNTFILE\" <<EOF\n");
+	fprintf(enforcer, "/\t\trx\n");
+	fprintf(enforcer, "/tmp\t$PWD/%s\trwx\n", tmp_path);
+	fprintf(enforcer, "/var/tmp\t$PWD/%s\trwx\n", vartmp_path);
+	fprintf(enforcer, "/dev/shm\t$PWD/%s\trwx\n", shm_path);
+	fprintf(enforcer, "/dev/null\trwx\n");
+	fprintf(enforcer, "/dev/zero\trwx\n");
+	fprintf(enforcer, "/dev/full\trwx\n");
+	fprintf(enforcer, "/dev/random\trwx\n");
+	fprintf(enforcer, "/dev/urandom\trwx\n");
+	fprintf(enforcer, "/home\t\tDENY\n");
+	list_first_item(input_list);
+	while((f=list_next_item(input_list))) {
+		fprintf(enforcer, "$PWD/%s\trwx\n", f->filename);
+	}
+	list_first_item(output_list);
+	while((f=list_next_item(output_list))) {
+		fprintf(enforcer, "$PWD/%s\trwx\n", f->filename);
+	}
+	fprintf(enforcer, "EOF\n\n");
+	fprintf(enforcer, "./parrot_run -m \"$MOUNTFILE\" -- \"$@\"\n");
+	fprintf(enforcer, "exit $?\n");
+	fclose(enforcer);
+
+	list_push_tail(enforcer_paths, dag_file_lookup_or_create(n->d, enforcer_path));
+
+	makeflow_log_file_expectation(n->d, enforcer_paths);
```

Comment:
From what I understood talking with Tim, the set of files in enforcer_paths are files that are created prior to the task being run, so outside the scope of any node and in the node prep step. We have two choices to log it properly, log expected when we make the list, then log exists when we create it, both prior to node submission, or simply log it as exists. The important idea that that prior to running we know the files exist, and then afterwards the node will look at these files and move them to completed. If they are in expected, we clean them in clean mode, but in GC we only look at completed files, so the files will never be moved to existing->completed->deleted despite us creating them.